Added a client side player name sanitation

Mailaender opened the pull request:
Closes #3200
Closes #3332
Closes #7739
Fixes #8031

Mailaender committed:
sanitize player names

pchote commented:
This sanitization will need to be done by the server when the client connects or changes their name.  If somebody sends a name with a miniyaml-breaking character then the client will crash before it gets a chance to sanitize it.

Phrohdoh commented on OpenRA.Game/Settings.cs:
```diff
@@ -300,6 +301,28 @@ public void Save()
+			var forbiddenNames = new string[] { "Open", "Closed", "Spectator" };
+
+			var clean = dirty;
+			if (string.IsNullOrEmpty(dirty) || forbiddenNames.Contains(clean) || clean.Contains(" AI"))
```
`Foobar AIBar` will be flagged as dirty.

pchote replied:
A simpler and more robust solution would be to check against the AI names directly.

pchote replied:
The TD and D2K bots don't have "AI" in their names.

Mailaender replied:
Right. Fixed.

Mailaender committed:
also sanitize server names

Mailaender commented:
I now also added a game server side sanitizing during client validation from the handshake and added a server name check as requested by @obrakmann. The rules are similar.

penev92 commented on OpenRA.Game/Settings.cs:
```diff
@@ -300,6 +302,52 @@ public void Save()
+
+			var clean = dirty;
+			if (string.IsNullOrEmpty(dirty))
+				clean = "";
```
This will probably need to have some non-empty default value (`Newbie` is the default player name, so probably that), otherwise the sanitation in this case doesn't really work.

Mailaender replied:
This will be replaced by `Newbie` and `OpenRA Game` (default server name) in the functions below.

Mailaender replied:
Wait a moment, you are right. My ordering of the checks is wrong. Entering `#` will allow an empty name.

penev92 commented on OpenRA.Game/Settings.cs:
```diff
@@ -300,6 +302,52 @@ public void Save()
+			else
+			{
+				foreach (var disallowedChar in disallowedChars)
+					clean = clean.Replace(disallowedChar.ToString(), string.Empty);
```
I'd suggest replacing with wither '-' or '_', because otherwise this can return an empty string, which leads us to my comment above.

Mailaender replied:
Doesn't matter. Empty strings will be replaced below. See my comment above.

Mailaender commented:
Optimized it again after the function split for server name checks.

deniz1a commented:
This changes LobbyUtils.cs at the same lines as #8118.

pchote commented:
The most important place to do the sanitization is LobbyLogic.IntepretCommand (the "name" handler).

A more polished client-side implementation would prevent the bad characters from being entered in the text fields in the first place, and then rely on the server-side implementation to deal with any hacked clients.

penev92 commented:
Also please don't forget:
>[20:11:21] < penev> SanitizedPlayerName() looks ok now
[20:11:33]	< penev> but I'm not convinced about the server one
[20:12:51]	< penev> because SanitizedPlayerName() sanitizes the name, and then checks if the new string is empty
[20:13:04]	< penev> SanitizedServerName() doesn't care if the sanitized name is empty
[20:13:33]	< penev> it checks before sanitizing
[20:13:46]	< penev> but if my server's name is []
[20:13:53]	< penev> it will do
[20:13:57]	< penev> return SanitizedName(dirty); 

... which will return an empty string.

Mailaender commented:
Well spotted. Updated.

abcdefg30 commented on OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:
```diff
@@ -717,9 +717,10 @@ public bool InterpretCommand(S server, Connection conn, Session.Client client, s
-						server.SendMessage("{0} is now known as {1}.".F(client.Name, s));
-						client.Name = s;
+						var sanitizedName = OpenRA.Settings.SanitizedPlayerName(s);
+						Log.Write("server", "Player@{0} is now known as {1}.", conn.Socket.RemoteEndPoint, sanitizedName);
```
Could you add a check if the name changed really?
If I try to change `abcdefg30` to `abcdefg30#` the `#` is removed and `abcdefg30 is now known as abcdefg30` appears.

abcdefg30 commented:
Otherwise :+1: / :white_check_mark:

Mailaender committed:
don't change player names without necessity

Mailaender commented:
Good idea. Done.

penev92 commented:
Looks good to me. Thanks! :+1:

penev92 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/e4a5f0000d2d7edaeabb4fc872b6b2b500cd1696...15337f861dae8f5e963ba309aba8d02665d87c98)
